fix: make lock-on client-sided

allvei opened the pull request:
closes #157

allvei committed:
fix: lock-on improvements
- Pass lock-on target selection is now adjustable and angle based, rather than distance from crosshair. Adjustable via `tf_passtime_lock_angle`.
- Made pass targeting client-authoritative by adding `passtime_lock_target` to the user command structure.
- Simplified eye-to-eye line of sight check by removing unnecessary matrix calculations.

allvei committed:
fix: disable server sided pass lock-on

drunderscore commented on src/game/shared/tf/passtime_convars.cpp:
```diff
@@ -12,81 +12,82 @@
 
 #define PASSTIME_CONVAR(NAME, STR, DESC) ConVar NAME(#NAME, #STR, FLAGS_DEFAULT, DESC)
 
-PASSTIME_CONVAR( tf_passtime_scores_per_round, 5, "Number of scores it takes to win a round. Similar to tf_flag_caps_per_round." );
```
One day I feel we ought to clang-format the world to avoid these kinds of diffs from sneaking in. (no action needed)

drunderscore commented on src/game/shared/tf/passtime_convars.h:
```diff
@@ -31,6 +31,7 @@ extern ConVar
 	tf_passtime_flinch_boost,
 	tf_passtime_mode_homing_lock_sec,
 	tf_passtime_mode_homing_speed,
+	tf_passtime_lock_angle,
```
This should be under `pf_`.

drunderscore commented on src/game/shared/tf/tf_weapon_passtime_gun.cpp:
```diff
@@ -552,15 +538,12 @@ void CPasstimeGun::ItemPostFrame()
-				filter.RemoveAllRecipients();
-				filter.AddRecipient( pCurrentTarget );
-				EmitSound( filter, pCurrentTarget->entindex(), kTargetHightlightSound );
+				pOwner->EmitSound( kTargetHightlightSound );
```
Emitting a sound is a side-effect, and needs to be protected with `prediction->IsFirstTimePredicted()` or similar to prevent the sound from being emitted multiple times if re-predicting (easily happens on some lag, test with `net_fakelag`)

drunderscore commented on src/game/shared/tf/tf_weapon_passtime_gun.cpp:
```diff
@@ -552,15 +538,12 @@ void CPasstimeGun::ItemPostFrame()
-				EmitSound( filter, pCurrentTarget->entindex(), kTargetHightlightSound );
+				pOwner->EmitSound( kTargetHightlightSound );
+				// play it for the target
+				pCurrentTarget->EmitSound( kTargetHightlightSound );
```
The original code was transcribed incorrectly here when swapping logical sides -- it doesn't make sense to play this sound on the carrier's client. This is causing _no sound_ to be heard when you are locked-on to.

drunderscore reviewed (dismissed):
I'm sure you're aware of the consequences of this being made client-authoritative. A cheat client can lock-on without LOS even which is slightly silly but possibly not a realistic concern (I specifically tested this).

Additionally, you may find `snd_show 1` to help with resolving some of the comments about the sound.

I used `net_fakelag 45` and `net_fakejitter 15` on a listen server for testing (meaning 90 ping minimum).

allvei committed:
Merge branch 'development' into allvei/issue157

Signed-off-by: xCape <mark.veiermann@gmail.com>

allvei committed:
fix: lock-on sound and convar prefix

- Added prediction->IsFirstTimePredicted() check to prevent duplicate sounds during re-prediction
- Restored proper target-only sound playback using CPASFilter instead of incorrectly playing on the carrier
- Renamed tf_passtime_lock_angle to p4ss_passtime_lock_angle

allvei commented:
Ignore this for now.

allvei committed:
fix: predict lock-on and unlock sound playback

- Added prediction check to lock sounds to prevent duplicate sounds during re-prediction
- Fixed lock sounds not playing for the target by moving them server side

allvei committed:
Merge branch 'development' into allvei/issue157

Signed-off-by: xCape <mark.veiermann@gmail.com>

allvei commented:
Tested with `net_fakelag 100` works good now, also no dupe sounds.
Not tested online.

CarlJosell commented on src/game/shared/tf/passtime_convars.h:
```diff
@@ -92,8 +92,8 @@ extern ConVar
 
```
p4ss_passtime_lock_angle needs to be in the header file aswell.

allvei replied:
You can make the change

allvei replied:
Resolved in latest commit.

CarlJosell reviewed (changes requested):
It seems like the lock angle convar was never initialized in the header file. Fails on build.

allvei committed:
Add missing p4ss_passtime_lock_angle in header

Signed-off-by: xCape <mark.veiermann@gmail.com>

allvei committed:
Merge branch 'development' into allvei/issue157

allvei committed:
Merge branch 'development' into allvei/issue157

CarlJosell reviewed (commented):
The actual issue is now fixed and works as intended. 

However there's an old issue with tf_hud_deathnotice where the function GetShortNick() doesn't return the correct type. Idk how this ended up here and why it's breaking now. 

Imo delete that piece of shit and get this PR in, and try to fix the shortname stuff for killfeed in another issue.

allvei committed:
Merge branch 'development' into allvei/issue157
